### PR TITLE
Allow fail_fast to be passed as a string

### PR DESCRIPTION
**Pull request by vfonic**
I thought I'll contribute a small PR with my little knowledge of ruby and rspec.

But I gotta say it was far from smooth sailing. :/

NOTE: I understand most of these are my own issues due to not reading documentation. (more on that towards the end)
The reason why I'm writing this lengthy explanation is to show how difficult it was for me to contribute to this project. I wanted to make relatively small change and have a discussion about it, but it looks like it will turn into a discussion about how to setup rspec-core for local development. :/

I'm willing, given time, to help with the documentation.

Here's my first time contributing to rspec-core story:

I tried to skip reading most of the documentation (SORRY!), assuming "standard" stuff will work (git clone, `bin/setup` and that should be it). Especially since I read that I don't have to clone all the repos.

So I cloned `rspec-core` and (stupidly) also decided to checkout `v3.8.0` tag. Then I ran `bundle install` and that failed as it was fetching `3.9.0.pre` from github.
Eventually I checkout `master` and `bundle install` passed. Then I ran `script/run_build`, and that tells me I need to create missing binstubs. I run `bundle binstubs rspec-core rake cucumber` and rerun `script/run_build`, but it failed with `cannot load such file -- rspec/support`:

<details>
<summary>Stacktrace</summary>

```
> script/run_build
============= Starting binstub check ===============
Checking required binstubs
============= Ending binstub check ===============
============= Starting specs ===============
/Users/viktor/Developer/Ruby/rspec-core/bin/rspec
Coverage report generated for RSpec to /Users/viktor/Developer/Ruby/rspec-core/coverage. 1 / 2 LOC (50.0%) covered.
Traceback (most recent call last):
	5: from script/rspec_with_simplecov:42:in `<main>'
	4: from script/rspec_with_simplecov:42:in `load'
	3: from /Users/viktor/Developer/Ruby/rspec-core/exe/rspec:3:in `<top (required)>'
	2: from /Users/viktor/Developer/Ruby/rspec-core/exe/rspec:3:in `require'
	1: from /Users/viktor/Developer/Ruby/rspec-core/lib/rspec/core.rb:5:in `<top (required)>'
/Users/viktor/Developer/Ruby/rspec-core/lib/rspec/core.rb:5:in `require': cannot load such file -- rspec/support (LoadError)
```

</details>

So I ran `script/clone_all_rspec_repos` and that finally allowed me to run `script/run_build`. But, I got failure that nokogiri gem failed to install (I lost that stacktrace). I installed nokogiri (`gem install nokogiri -v '1.8.5' --source 'https://rubygems.org/'`), and rerun `script/run_build`. Now I getting this error:

<details>
<summary>Stacktrace</summary>

```
> script/run_build
============= Starting binstub check ===============
Checking required binstubs
============= Ending binstub check ===============
============= Starting specs ===============
/Users/viktor/Developer/Ruby/rspec-core/bin/rspec

An error occurred while loading ./spec/rspec/core/memoized_helpers_spec.rb.
Hint: Install the `did_you_mean` gem in order to provide suggestions for similarly named files.
Failure/Error: require 'thread_order'

LoadError:
  cannot load such file -- thread_order
# ./spec/rspec/core/memoized_helpers_spec.rb:1:in `require'
# ./spec/rspec/core/memoized_helpers_spec.rb:1:in `<top (required)>'
# ./lib/rspec/core/configuration.rb:2053:in `load'
# ./lib/rspec/core/configuration.rb:2053:in `load_file_handling_errors'
# ./lib/rspec/core/configuration.rb:1560:in `block in load_spec_files'
# ./lib/rspec/core/configuration.rb:1558:in `each'
# ./lib/rspec/core/configuration.rb:1558:in `load_spec_files'
# ./lib/rspec/core/runner.rb:102:in `setup'
# ./lib/rspec/core/runner.rb:86:in `run'
# ./lib/rspec/core/runner.rb:71:in `run'
# ./lib/rspec/core/runner.rb:45:in `invoke'
# ./exe/rspec:4:in `<top (required)>'
# script/rspec_with_simplecov:42:in `load'
# script/rspec_with_simplecov:42:in `<main>'
Run options: exclude {:ruby=>#<Proc:./spec/spec_helper.rb:99>}


Top 0 slowest examples (0 seconds, 0.0% of total time):

Finished in 0.00006 seconds (files took 0.63136 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples

Coverage report generated for RSpec to /Users/viktor/Developer/Ruby/rspec-core/coverage. 2 / 2 LOC (100.0%) covered.
```

</details>

So, after trying `gem install thread_order` and still getting the same error, I decided to give up and just post what I came up with so far (as this PR) and check whether the tests on the CI will pass (locally, running `rspec` passes).

### CODE REVIEW

**Comment by vfonic** [comment]
### Why I made this PR in the first place?

I wanted to have my specs fail fast so I ran `FAIL_FAST=true rspec`, but that didn't...fail fast. :)

I then realized that ENV vars are passed in as strings and I had this in my `spec_helper.rb`:

```rb
config.fail_fast = ENV['FAIL_FAST'] || false
```

That obviously resolved to `'true'` (string) instead of `true` (boolean).

I wanted to check if anyone else did this, or where did I get this from. Then I [found](https://github.com/andyt/spree-magento-importer/blob/b3d36a7e0e18e532c1236faf4af8478757f81c47/spec/spec_helper.rb#L3) [numerous](https://github.com/cenit-io/cenit_cmd/blob/b9f3ac98c6e082b2c1b25d74c7ecb55d00ee1511/spec/spec_helper.rb#L17) [other](https://github.com/solidusio-contrib/solidus_countries_backend/blob/6ecf36e1e74c85c31c1c8a9220ea896a7a21b3da/spec/spec_helper.rb#L18) [repositories](https://github.com/chiaki0621/solidus/blob/978293d020d20895b99265ec1a92380b5d636ffc/vendor/bundle/ruby/2.4.0/gems/solidus_support-0.2.1/lib/solidus_support/extension/spec_helper.rb#L15) [that](https://github.com/Jamesecy/ubuntutest/blob/3f4b64f70c75fe43d2de2dca3efc160636b8bd1a/master_twilio_integration_worked/spree/cmd/spec/spec_helper.rb#L27) [did](https://github.com/nlandolfi/dawn-2/blob/55e0d5a34a7c931f3bd069a7371ba71cb82ef82f/spec/spec_helper.rb#L23) [exact](https://github.com/swattr/swattr/blob/dded12d6be116aab58a5c8f880c277e96675f576/spec/rails_helper.rb#L21) [same](https://github.com/seedchamp45/lampton/blob/f4696f5f67b9923e77214d673cdbea2c6d64da47/vendor/bundle/ruby/2.3.0/gems/solidus_support-0.2.0/lib/solidus_support/extension/spec_helper.rb#L15) [mistake](https://github.com/gxbsst/wineplus/blob/33f2a15c608ba604e28411d1e2aab5340f9dc0f9/lib/gems/spree/cmd/spec/spec_helper.rb#L26). So maybe it might be a good idea to allow passing a string and converting that string to a number of boolean if needed.

I think that `config.fail_fast = '3'` should follow the same logic like `config.fail_fast = 3` and that `config.fail_fast = 'true'` should follow the same logic like `config.fail_fast = true`.

I'm open to any suggestions on how to make the code better. `add_setting` seemed to be hard to extend to support type conversion so I added it only to `reporter` (which, when thinking about it, doesn't necessarily make sense that `reporter` is the one who stops execution, but also does type conversion.) I'm opet to suggestions. :)

File: lib/rspec/core/reporter.rb [hunk 1]
```diff
@@ -221,10 +221,10 @@ def abort_with(msg, exit_status)
     def fail_fast_limit_met?
       return false unless (fail_fast = @configuration.fail_fast)
 
-      if fail_fast == true
+      if fail_fast.to_s == 'true'
         @failed_examples.any?
       else
-        fail_fast <= @failed_examples.size
+        fail_fast.to_i <= @failed_examples.size
```
Comment (JonRowe) [inline lib/rspec/core/reporter.rb]: I'd prefer you handle both these edge cases in configuration.

File: spec/rspec/core/example_group_spec.rb [hunk 2]
```diff
@@ -1472,6 +1496,46 @@ def extract_execution_results(group)
...
+          group().before(:all) { raise "error in before all" }
+          group().example("equality") { expect(1).to eq(1) }
+          group().example("equality") { expect(1).to eq(1) }
+          group().example("equality") { expect(1).to eq(1) }
+
+          expect(group().run(reporter)).to be false
+
+          expect(RSpec.world.wants_to_quit).to be true
+        end
+      end
```
Comment (JonRowe) [inline spec/rspec/core/example_group_spec.rb]: All of these could then be collapsed into proving the configuration will always be `true` `false`, `n` or `nil`

**Comment by JonRowe** [comment]
TLDR I'm opening to making this change as it's a bit surprising, but only in configuration. You can change the setting to use `define_reader` then implement a custom writer to handle the conversion.

> I tried to skip reading most of the documentation (SORRY!), assuming "standard" stuff will work (git clone, bin/setup and that should be it). Especially since I read that I don't have to clone all the repos.

Hi, sorry you had trouble getting started, it is supposed to be a standard:

```
git clone ...
bundle install
rspec
```

> So I cloned rspec-core and (stupidly) also decided to checkout v3.8.0 tag. Then I ran bundle install and that failed as it was fetching 3.9.0.pre from github.

You will have to re run `bundle update` after changing versions, same with any gem.

> Then I ran script/run_build, and that tells me I need to create missing binstubs. I run bundle binstubs rspec-core rake cucumber and rerun script/run_build, but it failed with cannot load such file -- rspec/support:

If you checked our read me you'd see thats only for Travis and recreating Travis builds. Try just `rspec`

**Comment by vfonic** [comment]
Thanks for checking this out so quickly! I'll look into addressing your comments now.

**Comment by vfonic** [comment]
> You will have to re run bundle update after changing versions, same with any gem.

That's what I did to no avail:

<details>

```bash
> git clone rspec/rspec-core
> cd rspec-core
rspec-core > git checkout 3e1b5960cd4546378660bc1d84e5eaa74568ddef # sha of commit with v3.8.0 tag
rspec-core > git checkout -b v3.8.0
rspec-core > bundle install
Fetching https://github.com/rspec/rspec.git
Fetching https://github.com/rspec/rspec-expectations.git
Fetching https://github.com/rspec/rspec-mocks.git
Fetching https://github.com/rspec/rspec-support.git
The dependency jruby-openssl (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Bundler could not find compatible versions for gem "rspec-support":
  In Gemfile:
    rspec-support

    rspec-core was resolved to 3.8.0, which depends on
      rspec-support (~> 3.8.0)

    rspec-expectations was resolved to 3.9.0.pre, which depends on
      rspec-support (= 3.9.0.pre)


rspec-core > bundle update
Fetching https://github.com/rspec/rspec.git
Fetching https://github.com/rspec/rspec-expectations.git
Fetching https://github.com/rspec/rspec-mocks.git
Fetching https://github.com/rspec/rspec-support.git
The dependency jruby-openssl (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Bundler could not find compatible versions for gem "rspec-support":
  In Gemfile:
    rspec-support

    rspec-core was resolved to 3.8.0, which depends on
      rspec-support (~> 3.8.0)

    rspec-expectations was resolved to 3.9.0.pre, which depends on
      rspec-support (= 3.9.0.pre)
rspec-core 3.8.0 ✔ >
```

</details>



I actually checked out that part. Maybe the wording and order of bullet points create the confusion:

> ### Contributing
>
> * Build details
> * Code of Conduct
> * Detailed contributing guide
> * Development setup guide

Do you think it would make sense to move "Development setup guide" more up?

Then I also read this:
https://github.com/rspec/rspec-core/blob/master/BUILD_DETAIL.md#the-ci-build-in-detail

(this is the first bullet point under "Contributing")

Then I also saw this:

https://github.com/rspec/rspec-core/blob/master/DEVELOPMENT.md#running-the-build

This is under "Development setup guide". No mention of "just run `rspec` locally".

It made me believe the correct thing to do is to "run the build locally" before submitting the PR so I tried the best I can to make that happen.

File: lib/rspec/core/configuration.rb [hunk 3]
```diff
@@ -202,10 +202,31 @@ def only_failures_but_not_configured?
...
-      # up and exit (default: `nil`).
-      add_setting :fail_fast
+      # up and exit (default: `nil`). Can also be `true` to fail on exit on first
+      # failure
+      define_reader :fail_fast
+
+      # @see fail_fast
+      def fail_fast=(value)
+        case value
+        when true, 'true'
```
Comment (vfonic) [inline lib/rspec/core/configuration.rb]: Let me know if you wanted to cover more truthy values. I figured this will cover most use-cases.

File: lib/rspec/core/configuration.rb [hunk 4]
```diff
@@ -202,10 +202,31 @@ def only_failures_but_not_configured?
...
+          @fail_fast = true
+        when false, 'false'
+          @fail_fast = false
+        when nil
+          @fail_fast = nil
+        else
+          if value.to_i == 0
+            raise ArgumentError, "Cannot set `RSpec.configuration.fail_fast`" \
```
Comment (vfonic) [inline lib/rspec/core/configuration.rb]: I decided to throw here so that the users can know when they set this value incorrectly, and not have to wonder why the configuration is not working.
Would this then require a major version bump?

Do you prefer not to have this config throw?

EDIT: We could probably just issue a warning?

File: lib/rspec/core/configuration.rb (hunk 3 above)
Comment (JonRowe) [reply lib/rspec/core/configuration.rb]: No this is fine :)

File: lib/rspec/core/configuration.rb (hunk 4 above)
Comment (JonRowe) [reply lib/rspec/core/configuration.rb]: I'm on the fence, if the previous behaviour is considered truthy then a warning should be issued and it considered truthy, with a note to deprecate and raise in RSpec 4, but otherwise I consider this a bug fix and it can be raised.

File: lib/rspec/core/configuration.rb [hunk 5]
```diff
@@ -399,6 +420,7 @@ def treat_symbols_as_metadata_keys_with_true_values=(_value)
       # @overload shared_context_metadata_behavior=(value)
       #   @param value [:trigger_inclusion, :apply_to_host_groups] sets the configured behavior
       define_reader :shared_context_metadata_behavior
+
```
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: Please don't add random whitespace :)

File: lib/rspec/core/configuration.rb (hunk 4 above)
Comment (vfonic) [reply lib/rspec/core/configuration.rb]: 🤔

Thinking more about this, I'd lean towards warning + raise in rspec 4, so that users don't get failing rspec runner, for a non-major version update.

I'll go ahead and change to warning. I think it will create less frustration than raising an error.

File: lib/rspec/core/configuration.rb (hunk 5 above)
Comment (vfonic) [reply lib/rspec/core/configuration.rb]: I noticed there's always an empty line before `define_reader :setting` and `def setting=(value)`, except for this place here.

Since the comment in there is:
```ruby
      define_reader :shared_context_metadata_behavior
      # @see shared_context_metadata_behavior
      def shared_context_metadata_behavior=(value)
```

I figured it would be better to add this empty line, to be consistent. I'll revert this, it's not related to the rest of the changes.

File: lib/rspec/core/configuration.rb (hunk 4 above)
Comment (vfonic) [reply lib/rspec/core/configuration.rb]: I found three different ways the warnings are being set in this file:

```ruby
warn "RSpec's reporter has already been initialized with " \
  "#{deprecation_stream.inspect} as the deprecation stream, so your change to "\
  "`deprecation_stream` will be ignored. You should configure it earlier for " \
  "it to take effect, or use the `--deprecation-out` CLI option. " \
  "(Called from #{CallerFilter.first_non_rspec_line})"

RSpec.warning "Configuring `#{name}` to #{value} has no effect since " \
              "RSpec has already loaded the spec files."

RSpec.warn_with "WARNING: `:suite` hooks do not support metadata since " \
                "they apply to the suite as a whole rather than " \
                "any individual example or example group that has metadata. " \
                "The metadata you have provided (#{meta.inspect}) will be ignored."
```

I went for using `warn`.

Comment (JonRowe) [reply lib/rspec/core/configuration.rb]: Can you use one of our helpers please, not Rubys bare warn. It provides a configurable suppression point.

File: lib/rspec/core/configuration.rb [hunk 6]
```diff
@@ -202,10 +202,33 @@ def only_failures_but_not_configured?
...
+      # up and exit (default: `nil`). Can also be `true` to fail and exit on first
+      # failure
+      define_reader :fail_fast
+
+      # @see fail_fast
+      def fail_fast=(value)
+        case value
+        when true, 'true'
+          @fail_fast = true
+        when false, 'false'
```
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: Add `0` here

File: lib/rspec/core/configuration.rb [hunk 7]
```diff
@@ -202,10 +202,33 @@ def only_failures_but_not_configured?
...
+          @fail_fast = false
+        when nil
+          @fail_fast = nil
+        else
+          @fail_fast = value.to_i
+
+          if value.to_i == 0
+            # TODO: in RSpec 4, consider raising an error here.
+            warn "Cannot set `RSpec.configuration.fail_fast`" \
```
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: Can you use our warning helper here.

File: lib/rspec/core/configuration.rb (hunk 4 above)
Comment (vfonic) [reply lib/rspec/core/configuration.rb]: Which one would you recommend? I'll switch to `RSpec.warning`

Comment (JonRowe) [reply lib/rspec/core/configuration.rb]: Either is fine, warning adds "WARNING" to the front if you don't want to type it yourself :joy:

Comment (vfonic) [reply lib/rspec/core/configuration.rb]: Just to note, the original behavior here is an error raised in rspec code due to this line:
https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/reporter.rb#L227

```
Failure/Error: fail_fast <= @failed_examples.size

ArgumentError:
  comparison of String with 1 failed
```

This makes the test fail, but continues executing the following tests. The following tests all fail, no matter whether they would pass in isolation (because `config.fail_fast` gets checked in reporter on every test I guess).

**Comment by benoittgt** [comment]
Thanks for this PR @vfonic 

> Do you think it would make sense to move "Development setup guide" more up?

I think it's a good idea :)

File: lib/rspec/core/configuration.rb [hunk 8]
```diff
@@ -202,10 +202,33 @@ def only_failures_but_not_configured?
...
+        when nil
+          @fail_fast = nil
+        else
+          @fail_fast = value.to_i
+
+          if value.to_i == 0
+            # TODO: in RSpec 4, consider raising an error here.
+            RSpec.warning "Cannot set `RSpec.configuration.fail_fast`" \
+              " to `#{value.inspect}`. Only `true`, `false`, `nil` and integers" \
+              " are valid values."
```
Comment (benoittgt) [inline lib/rspec/core/configuration.rb]: But I don't understand. If you provide a 0, it is an integer so it should valid and you should not receive the warning?

Do I miss something? Or should validate only positive integers?

Comment (JonRowe) [reply lib/rspec/core/configuration.rb]: `0` is the same as turning it off

Comment (benoittgt) [reply lib/rspec/core/configuration.rb]: Yes but if you turn it off with 0 it is still a valid integer?

This not a blocker for me to validate this PR 🙂

Comment (vfonic) [reply lib/rspec/core/configuration.rb]: 🤔 Indeed. We could be more specific by saying

> only `true`, `false`, `nil` and positive integers are valid values

I'm not sure if we should be that specific. Keep in mind that, with this PR, we treat `'true'` (string) the same as `true`, yet, it's not mentioned in the warning message above. (and I don't think it should be explicitly mentioned)

Comment (benoittgt) [reply lib/rspec/core/configuration.rb]: 😄 I think it is good enough

**Comment by vfonic** [comment]
Hey @JonRowe are you still waiting for some changes to be made?

**Comment by JonRowe** [comment]
Thanks!
